In [23]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt


In [34]:
# Loading the datasets
data_2022 = pd.read_csv('/content/PA_l2_2022stats_county.csv')
data_2020 = pd.read_csv('/content/PA_l2_2022stats_2020block.csv')

# Displaying the first 5 rows of the dataset
display(data_2022.head())
display(data_2020.head())

<ipython-input-34-39a9a2369675>:3: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  data_2020 = pd.read_csv('/content/PA_l2_2022stats_2020block.csv')


,countyfips,countyname,total_reg,age_18_19,age_20_24,age_25_29,age_30_34,age_35_44,age_45_54,age_55_64,...,s20200114_pct_voted_aa,s20200114_voted_esa,s20200114_reg_esa,s20200114_pct_voted_esa,s20200114_voted_oth,s20200114_reg_oth,s20200114_pct_voted_oth,s20200114_voted_unk,s20200114_reg_unk,s20200114_pct_voted_unk
0,42011,BERKS,250183,1522,15637,18149,19213,38748,39055,47124,...,0.0,0,2315,0.0,1,3042,0.0003,2,23324,0.0001
1,42013,BLAIR,75245,427,4312,4762,5158,11322,11640,13245,...,0.0,0,350,0.0,0,833,0.0000,0,7463,0.0000
2,42001,ADAMS,67016,487,4322,4600,4412,8881,9539,12837,...,0.0,0,368,0.0,0,529,0.0000,3,5761,0.0005
3,42003,ALLEGHENY,871081,5727,55433,79281,85346,147259,118718,138308,...,0.0,0,17067,0.0,0,23460,0.0000,0,103480,0.0000
4,42005,ARMSTRONG,39649,222,2065,2311,2496,5438,6117,7672,...,0.0,0,186,0.0,0,564,0.0000,0,4615,0.0000


,geoid20,total_reg,age_18_19,age_20_24,age_25_29,age_30_34,age_35_44,age_45_54,age_55_64,age_65_74,...,s20200114_pct_voted_aa,s20200114_voted_esa,s20200114_reg_esa,s20200114_pct_voted_esa,s20200114_voted_oth,s20200114_reg_oth,s20200114_pct_voted_oth,s20200114_voted_unk,s20200114_reg_unk,s20200114_pct_voted_unk
0,421330240022036,6,0,0,1,1,0,2,1,0,...,NaN,0,0,NaN,0,0,NaN,0,0,NaN
1,421330240022035,48,0,2,3,6,6,9,8,10,...,NaN,0,1,0.0,0,0,NaN,1,4,0.25
2,421330240022034,0,0,0,0,0,0,0,0,0,...,NaN,0,0,NaN,0,0,NaN,0,0,NaN
3,421330240022033,2,0,1,1,0,0,0,0,0,...,NaN,0,0,NaN,0,0,NaN,0,0,NaN
4,421330240022032,30,0,0,6,4,1,5,7,4,...,NaN,0,1,0.0,0,0,NaN,0,5,0.00


In [36]:
# Printing the column names
print(data_2022.columns)
print(data_2020.columns)

Index(['countyfips', 'countyname', 'total_reg', 'age_18_19', 'age_20_24',
       'age_25_29', 'age_30_34', 'age_35_44', 'age_45_54', 'age_55_64',
       ...
       's20200114_pct_voted_aa', 's20200114_voted_esa', 's20200114_reg_esa',
       's20200114_pct_voted_esa', 's20200114_voted_oth', 's20200114_reg_oth',
       's20200114_pct_voted_oth', 's20200114_voted_unk', 's20200114_reg_unk',
       's20200114_pct_voted_unk'],
      dtype='object', length=399)
Index(['geoid20', 'total_reg', 'age_18_19', 'age_20_24', 'age_25_29',
       'age_30_34', 'age_35_44', 'age_45_54', 'age_55_64', 'age_65_74',
       ...
       's20200114_pct_voted_aa', 's20200114_voted_esa', 's20200114_reg_esa',
       's20200114_pct_voted_esa', 's20200114_voted_oth', 's20200114_reg_oth',
       's20200114_pct_voted_oth', 's20200114_voted_unk', 's20200114_reg_unk',
       's20200114_pct_voted_unk'],
      dtype='object', length=398)


In [37]:
# Checking for missing values
data_2022.dropna(inplace=True)
data_2020.dropna(inplace=True)

display(data_2022.head())
display(data_2020.head())

,countyfips,countyname,total_reg,age_18_19,age_20_24,age_25_29,age_30_34,age_35_44,age_45_54,age_55_64,...,s20200114_pct_voted_aa,s20200114_voted_esa,s20200114_reg_esa,s20200114_pct_voted_esa,s20200114_voted_oth,s20200114_reg_oth,s20200114_pct_voted_oth,s20200114_voted_unk,s20200114_reg_unk,s20200114_pct_voted_unk
0,42011,BERKS,250183,1522,15637,18149,19213,38748,39055,47124,...,0.0,0,2315,0.0,1,3042,0.0003,2,23324,0.0001
1,42013,BLAIR,75245,427,4312,4762,5158,11322,11640,13245,...,0.0,0,350,0.0,0,833,0.0000,0,7463,0.0000
2,42001,ADAMS,67016,487,4322,4600,4412,8881,9539,12837,...,0.0,0,368,0.0,0,529,0.0000,3,5761,0.0005
3,42003,ALLEGHENY,871081,5727,55433,79281,85346,147259,118718,138308,...,0.0,0,17067,0.0,0,23460,0.0000,0,103480,0.0000
4,42005,ARMSTRONG,39649,222,2065,2311,2496,5438,6117,7672,...,0.0,0,186,0.0,0,564,0.0000,0,4615,0.0000


,geoid20,total_reg,age_18_19,age_20_24,age_25_29,age_30_34,age_35_44,age_45_54,age_55_64,age_65_74,...,s20200114_pct_voted_aa,s20200114_voted_esa,s20200114_reg_esa,s20200114_pct_voted_esa,s20200114_voted_oth,s20200114_reg_oth,s20200114_pct_voted_oth,s20200114_voted_unk,s20200114_reg_unk,s20200114_pct_voted_unk
647,421330238241023,136,0,3,1,28,40,20,17,15,...,0.0,0,6,0.0,0,1,0.0,0,20,0.0
2116,421330233013021,74,0,5,5,4,15,18,14,7,...,0.0,0,2,0.0,0,1,0.0,0,7,0.0
2158,421330233012014,74,0,9,7,2,25,18,8,3,...,0.0,0,3,0.0,0,1,0.0,0,7,0.0
2168,421330233012004,47,0,5,3,4,12,17,5,1,...,0.0,0,3,0.0,0,1,0.0,0,1,0.0
2199,421330233011000,305,1,20,12,17,33,47,48,67,...,0.0,0,7,0.0,0,8,0.0,0,30,0.0


In [38]:
# Finding the maximum votes received by each party in each row
data_2022['max_votes'] = data_2022[['party_dem', 'party_rep', 'party_npp', 'party_rin', 'party_lib', 'party_grn', 'party_con', 'party_ind', 'party_oth', 'party_unk']].max(axis=1)

# Calculating the win margins for each party as the difference from the max votes
data_2022['margin_dem'] = data_2022['max_votes'] - data_2022['party_dem']
data_2022['margin_rep'] = data_2022['max_votes'] - data_2022['party_rep']
data_2022['margin_npp'] = data_2022['max_votes'] - data_2022['party_npp']
data_2022['margin_rin'] = data_2022['max_votes'] - data_2022['party_rin']
data_2022['margin_lib'] = data_2022['max_votes'] - data_2022['party_lib']
data_2022['margin_grn'] = data_2022['max_votes'] - data_2022['party_grn']
data_2022['margin_con'] = data_2022['max_votes'] - data_2022['party_con']
data_2022['margin_ind'] = data_2022['max_votes'] - data_2022['party_ind']
data_2022['margin_oth'] = data_2022['max_votes'] - data_2022['party_oth']
data_2022['margin_unk'] = data_2022['max_votes'] - data_2022['party_unk']

# Displaying the data with win margins
display(data_2022[['countyname', 'margin_dem', 'margin_rep', 'margin_npp', 'margin_rin', 'margin_lib', 'margin_grn', 'margin_con', 'margin_ind', 'margin_oth', 'margin_unk']])


,countyname,margin_dem,margin_rep,margin_npp,margin_rin,margin_lib,margin_grn,margin_con,margin_ind,margin_oth,margin_unk
0,BERKS,720,0,72475,103093,103947,105057,105392,105404,102464,105415
1,BLAIR,26756,0,38891,45420,45990,46281,46341,46345,45881,46350
2,ADAMS,20843,0,30292,37916,38351,38649,38712,38720,37999,38722
3,ALLEGHENY,0,250860,394564,486942,492878,496213,497196,497158,488556,497272
4,ARMSTRONG,13217,0,20962,23727,24069,24240,24257,24265,24026,24268
...,...,...,...,...,...,...,...,...,...,...,...
62,SNYDER,9965,0,12611,14235,14402,14496,14512,14516,14326,14518
63,VENANGO,9950,0,15286,17853,18186,18347,18370,18372,18250,18372
64,WASHINGTON,9040,0,51254,62991,63789,64301,64385,64360,63282,64409
65,WESTMORELAND,27341,0,95436,114858,116431,117555,117676,117692,115749,117710


In [56]:
# Saving DataFrame with win margins to a CSV file
data_2022[['countyname', 'margin_dem', 'margin_rep', 'margin_npp', 'margin_rin', 'margin_lib', 'margin_grn', 'margin_con', 'margin_ind', 'margin_oth', 'margin_unk']].to_csv('County_Margins_2022.csv', index=False)


EXPLANATION:
Where the party shows a margin of '0', it means that the party won the elections in that county and the others with high numbers is the difference of votes they have with the winning party with '0'

EX: For county: BERKS, the republicans won the elections in that county and the democrats had 720 fewer votes.

In [44]:
# Aggregating the sum of votes for each party in 2020 and 2022
sum_2020 = data_2020[[
    'party_dem', 'party_rep', 'party_npp', 'party_rin',
    'party_lib', 'party_grn', 'party_con', 'party_ind',
    'party_oth', 'party_unk'
]].sum()

sum_2022 = data_2022[[
    'party_dem', 'party_rep', 'party_npp', 'party_rin',
    'party_lib', 'party_grn', 'party_con', 'party_ind',
    'party_oth', 'party_unk'
]].sum()

# Creating a DataFrame to display the results
results_df = pd.DataFrame({
    'Party': sum_2020.index,
    'Votes_2020': sum_2020.values,
    'Votes_2022': sum_2022.values
})

# Calculating the difference
results_df['Difference'] = results_df['Votes_2022'] - results_df['Votes_2020']

# Displaying the DataFrame
display(results_df)

,Party,Votes_2020,Votes_2022,Difference
0,party_dem,417005,3744709,3327704
1,party_rep,223724,3297920,3074196
2,party_npp,105828,978267,872439
3,party_rin,8043,92289,84246
4,party_lib,3479,40049,36570
5,party_grn,911,9156,8245
6,party_con,56,906,850
7,party_ind,71,740,669
8,party_oth,10816,85100,74284
9,party_unk,0,0,0


In [60]:
# Saving the DataFrame to a CSV file
results_df.to_csv('Shifts_Since_2020.csv', index=False)

EXPLANATION: These are the changes in the number of registered voters for each political party between the 2020 and 2022 elections.